src/Compilers/CSharp/Portable/BoundTree/BoundNodeExtensions.cs

```diff
@@ -94,5 +97,153 @@ private class ContainsAwaitVisitor : BoundTreeWalkerWithStackGuardWithoutRecursi
+        /// </param>
+        /// <param name="binaryOperatorFactory">
+        /// Rewriter for the BoundBinaryOperator parts fo the binary operator. Passed the callback parameter, the original binary operator, and
+        /// the rewritten left and right components.
+        /// </param>
+        public static TResult RewriteInterpolatedStringAddition<TInterpolatedStringType, TArg, TResult>(
+            this BoundBinaryOperator binary,
+            TArg arg,
+            Func<TInterpolatedStringType, int, TArg, TResult> interpolatedStringFactory,
+            Func<BoundBinaryOperator, TResult, TResult, TArg, TResult> binaryOperatorFactory)
+            where TInterpolatedStringType : BoundInterpolatedStringBase
+        {
+            int i = 0;
+            var originalStack = ArrayBuilder<BoundBinaryOperator>.GetInstance();
+
+            var result = doRewrite(binary, arg, interpolatedStringFactory, binaryOperatorFactory, originalStack, ref i);
+
+            originalStack.Free();
+            return result;
+
+            static TResult doRewrite(
+                BoundBinaryOperator binary,
+                TArg arg,
+                Func<TInterpolatedStringType, int, TArg, TResult> interpolatedStringFactory,
+                Func<BoundBinaryOperator, TResult, TResult, TArg, TResult> binaryOperatorFactory,
+                ArrayBuilder<BoundBinaryOperator> originalStack,
+                ref int i)
+            {
+                TResult? result = default;
+
+                pushLeftNodes(binary, originalStack);
+
+                while (originalStack.TryPop(out var currentBinary))
+                {
+                    Debug.Assert(currentBinary.Left is TInterpolatedStringType || result != null);
+                    TResult rewrittenLeft = currentBinary.Left switch
+                    {
+                        TInterpolatedStringType interpolatedString => interpolatedStringFactory(interpolatedString, i++, arg),
+                        BoundBinaryOperator => result!,
+                        _ => throw ExceptionUtilities.UnexpectedValue(currentBinary.Left.Kind)
+                    };
+
+                    // For simplicity, we use recursion for binary operators on the right side of the tree. We're not traditionally concerned
+                    // with long chains of operators on the right side, as without parentheses we'll naturally make a tree that is deep on the
+                    // left side. If this ever changes, we can make this algorithm a more complex post-order iterative rewrite instead.
+
+                    var rewrittenRight = currentBinary.Right switch
+                    {
+                        TInterpolatedStringType interpolatedString => interpolatedStringFactory(interpolatedString, i++, arg),
+                        BoundBinaryOperator binaryOperator => doRewrite(binaryOperator, arg, interpolatedStringFactory, binaryOperatorFactory, originalStack, ref i),
```

Comment:
Re-using the same `originalStack` here feels dangerous/wrong, as `doRewrite` eventually returns the rewritten version of the last node on the stack, but in this recursive case that's not the one we want.